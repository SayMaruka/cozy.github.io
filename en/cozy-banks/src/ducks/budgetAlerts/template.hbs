
{{#extend "bank-layout"}}
  {{#content "emailTitle"}}
    {{t 'Notifications.categoryBudgets.email.innerTitle'}}
  {{/content}}
  {{#content "emailSubtitle"}}
    {{ formatDate date }}
  {{/content}}

  {{#content "content"}}
    <mj-section padding-top="0" padding-bottom="10px">
      <mj-column>
        <mj-text padding-top="0" padding-bottom="0">
          <h3 class='js-institution'>
            {{ budgetAlert.accountOrGroupLabel }}
          </h3>
        </mj-text>
        {{#each budgetAlerts as |budgetAlert| }}
          <mj-table>
            <tr class="js-transaction tableRow">
              <td width="40px">
                {{ categoryIcon budgetAlert.categoryId }}
              </td>
              <td>
                <mj-text>
                  {{ budgetAlert.categoryLabel }}
                </mj-text>
              </td>
              <td width="150px" align='right'>
                {{ budgetAlert.currentAmount }}€ {{ t "Notifications.categoryBudgets.email.budgetOutOf" }} {{ budgetAlert.maxThreshold }}€
              </td>
            </tr>
          </mj-table>
        {{/each}}
      </mj-column>
    </mj-section
  {{/content}}
{{/extend}}
