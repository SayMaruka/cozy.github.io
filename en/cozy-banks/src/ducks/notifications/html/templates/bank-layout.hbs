{{#extend "cozy-layout"}}
  {{#content "logoUrl"}}https://downcloud.cozycloud.cc/upload/cozy-banks/email-assets/logo-bank.png{{/content}}

  {{#content "appName"}}
    Banks
  {{/content}}

  {{#content "topLogo"}}<mj-image width="100px" src="https://downcloud.cozycloud.cc/upload/cozy-banks/email-assets/logo-cozy.png" />{{/content}}

  {{#content "appURL"}}{{ banksUrl }}{{/content}}

{{/extend}}
