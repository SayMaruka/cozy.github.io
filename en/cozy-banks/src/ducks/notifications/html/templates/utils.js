const Handlebars = require('handlebars')
const overEvery = require('lodash/overEvery')
const fromPairs = require('lodash/fromPairs')
const layouts = require('handlebars-layouts')

layouts.register(Handlebars)

const isOfType = type => node => {
  return node.type == type
}

const isPath = path => node => {
  return node.path.original == path
}

const isExtendBlock = overEvery([isOfType('BlockStatement'), isPath('extend')])
const isContentBlock = overEvery([
  isOfType('BlockStatement'),
  isPath('content')
])

/**
 * Extracts the name of extending template.
 * Extracts the AST of each content block.
 */
const extractInfo = templateContent => {
  const ast = Handlebars.parse(templateContent)
  const extendNode = ast.body.find(isExtendBlock)
  const contentNodes = extendNode.program.body.filter(isContentBlock)

  // Mapping from content block name to content block AST
  const contents = fromPairs(
    contentNodes.map(node => [node.params[0].value, node.program])
  )
  return {
    parentTemplate: extendNode.params[0].value,
    parts: contents,
    ast
  }
}

/**
 * Is used to replace parts of a template by rendered content.
 *
 * @param {Object} blocks - Block name -> Block AST
 * @param {Object} replacements - Block name -> Rendered content
 *
 * @example
 * ```
 * {{#content "emailSubtitle"}}
 *   {{#each transactions as |transaction| }}
 *     <div>{{ transaction.label }}</div>
 *   {{/each}}
 * {{/content}}
 *
 * becomes
 *
 * {{#content "emailSubtitle"}}
 *   <div>{{ transaction.label }}</div>
 *   <div>{{ transaction.label }}</div>
 * {{/content}}
 * ```
 */
const replaceParts = (blocks, replacements) => {
  for (const [blockName, replacement] of Object.entries(replacements)) {
    blocks[blockName].body = [
      {
        type: 'ContentStatement',
        original: replacement,
        value: replacement
      }
    ]
  }
}

module.exports = {
  extractInfo,
  replaceParts
}
