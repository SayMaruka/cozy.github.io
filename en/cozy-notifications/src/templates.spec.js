import appTemplate from './__tests__/app-layout.hbs'
import template from './__tests__/email-layout.hbs'
import { renderer } from './templates'

const helpers = {
  greeting: name => `Hello ${name} !`,
  t: x => x
}

const partials = {
  'app-layout': appTemplate
}

describe('template utils', () => {
  it('should extract parts and parents from template', () => {
    const { collectInfo } = renderer({ partials, helpers })
    const templateInfo = collectInfo(template)
    expect(templateInfo.parents).toEqual(['app-layout', 'cozy-layout'])
    expect(Object.keys(templateInfo.contentASTByName)).toEqual([
      'emailTitle',
      'emailSubtitle',
      'content',
      'logoUrl',
      'appName',
      'topLogo',
      'footerHelp'
    ])
  })

  it('should inject content blocks', () => {
    const { collectInfo, injectContent, Handlebars } = renderer({
      partials,
      helpers
    })
    const { ast } = collectInfo(template, partials)
    injectContent(ast, {
      emailTitle: 'emailTitle',
      emailSubtitle: 'emailSubtitle',
      content: 'content',
      logoUrl: 'logoUrl',
      appName: 'appName',
      topLogo: 'topLogo',
      appURL: 'appURL',
      settingsURL: 'settingsUrl',
      footerHelp: 'footerHelp'
    })
    const compiled = Handlebars.compile(ast)
    const rendered = compiled()
    expect(rendered).toMatchSnapshot()
  })

  it('should render parts and full', () => {
    const { render } = renderer({ partials, helpers })
    const data = {
      name: 'Homer',
      appUrl: 'https://homer-myapp.mycozy.cloud',
      settingsUrl: 'https://homer-settings.mycozy.cloud'
    }
    const compiled = render({
      template,
      data
    })
    expect(compiled.full).toMatchSnapshot()
    expect(compiled.parts.emailSubtitle.trim()).toBe('Hello Homer !')
  })
})
