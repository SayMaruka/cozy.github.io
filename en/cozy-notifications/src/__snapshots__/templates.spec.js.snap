// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template utils should inject content blocks 1`] = `
"<mjml>
   <mj-head>
      <mj-font name=\\"Lato\\" href=\\"https://fonts.googleapis.com/css?family=Lato:400,700\\" />
      <mj-style inline=\\"inline\\">
        h1, h2, h3, h4, h5, h6 {
  display:block;
  margin:0;
  padding:0;
}

/* header */
.header__title a {
  text-decoration: none;
  color: inherit;
  white-space: nowrap;
}

.header__desc {
  white-space: nowrap;
  text-align: right;
}
.header__desc__title {
  color: #32363F;
  font-weight: bold;
  font-size: 16px;
  margin-bottom: 4px;
  letter-spacing: 0.5px;
}
.header__desc__subtitle {
  color: #95999D;
  font-size: 14px;
  letter-spacing: 0.4px;
}

.account h3 {
  font-size: 12px;
  font-weight: bold;
  color: #95999D;
  text-transform: uppercase;
  letter-spacing: 0.6px;
}
.account h4 {
  font-size: 12px;
  font-weight: normal;
  color: #95999D;
  text-transform: uppercase;
  letter-spacing: 0.6px;
}
/* amount */
.amount {
  white-space: nowrap;
  font-size: 16px;
}
.amountPos {
  color: #35CE68;
}
.amountNeg {
  color: #F52D2D;
}
/* footer */
.footer__settings > td > div,
.footer__help > td > div {
  font-size: 12px;
  letter-spacing: 0.6px;
}

.footer__settings > td > div  {
  color: #32363F !important;
}
.footer__help > td > div {
  color: #95999D !important;
  text-align: right !important;
}
.footer__help a {
  display: inline-block;
  color: #95999D !important;
}

.footer__help a + a {
  margin-left: 16px;
}
/* text */
.txtRight {
  text-align: right;
}

/* media */
table {
  border-spacing: 0;
}

.tableRow {
  border-bottom: 1px #D6D8Da solid;
  padding: 8px 0;
  line-height: 1.5;
  letter-spacing: 0.25px;
}

.tableRow {
  border-bottom: 0;
}

.tableRow a, .tableRow a:visited {
  color: inherit;
  text-decoration: none;
}

@media screen and (max-width: 480px) {
  .header__desc__title {
    text-align: center !important;
  }
  .header__desc__subtitle {
    text-align: center !important;
  }
  .header__title > td {
    text-align: center !important;
  }
  .header__title > td > table {
    width: 100px;
    margin: auto;
  }

  .footer__settings > td > div {
    text-align: center !important;
  }

  .footer__help > td > div,
  td.footer__help > div,
  emailHeader {
    text-align: center !important;
  }
  .header__title table {
    margin: 0 auto;
    text-align: center;
  }
}

      </mj-style>
      <mj-attributes>
        <mj-all font-family=\\"Lato, Helvetica, Arial, sans-serif\\"></mj-all>
      </mj-attributes>
   </mj-head>
  <mj-body background-color=\\"#f5f6f7\\">
    <mj-section padding-top=\\"0\\" padding-bottom=\\"0\\">
      <mj-column>
        topLogo
      </mj-column>
    </mj-section>

    <!-- white sheet -->
    <mj-wrapper background-color=\\"white\\" border='1px solid #d6d6d8' border-top=\\"none\\">
      <!-- header -->
      <mj-section padding-bottom=\\"0\\" padding-top=\\"0\\">
        <mj-column>
          <mj-table width=\\"100%\\" css-class=\\"header__title\\" cellpadding='4px'>
            <tr>
              <td width=\\"30px\\">
                <a href=\\"appURL\\">
                  <img align=\\"left\\" width=\\"30px\\" src=\\"logoUrl\\" alt=\\"appName\\" />
                </a>
              </td>
              <td>
                <h1 style=\\"line-height: 32px\\"><a href=\\"appURL\\">appName</a></h1>
              </td>
            </tr>
          </mj-table>
        </mj-column>
        <mj-column css-class=\\"emailHeader\\">
          <mj-text padding-top=\\"0\\" padding-bottom=\\"0\\">
              <div class=\\"header__desc__title txtRight\\">
                emailTitle
              </div>
              <div class=\\"header__desc__subtitle txtRight\\">
                emailSubtitle
              </div>
          </mj-text>
        </mj-column>
      </mj-section>

      <mj-section css-class='before-content' padding-top=\\"8px\\" padding-bottom=\\"8px\\">
        <mj-column>
          <mj-divider border-width=\\"1px\\" border-color=\\"#95999d\\" />
        </mj-column>
      </mj-section>


      <!-- content -->
      content
    </mj-wrapper>

    <!-- footer -->
    <mj-section padding-top=\\"0\\">
      <mj-column>
        
      </mj-column>
      <mj-column>
        <mj-text align='right' css-class=\\"footer__help\\">
          footerHelp
        </mj-text>
      </mj-column>
    </mj-section>
  </mj-body>
</mjml>



"
`;

exports[`template utils should render parts and full 1`] = `
"<mjml>
   <mj-head>
      <mj-font name=\\"Lato\\" href=\\"https://fonts.googleapis.com/css?family=Lato:400,700\\" />
      <mj-style inline=\\"inline\\">
        h1, h2, h3, h4, h5, h6 {
  display:block;
  margin:0;
  padding:0;
}

/* header */
.header__title a {
  text-decoration: none;
  color: inherit;
  white-space: nowrap;
}

.header__desc {
  white-space: nowrap;
  text-align: right;
}
.header__desc__title {
  color: #32363F;
  font-weight: bold;
  font-size: 16px;
  margin-bottom: 4px;
  letter-spacing: 0.5px;
}
.header__desc__subtitle {
  color: #95999D;
  font-size: 14px;
  letter-spacing: 0.4px;
}

.account h3 {
  font-size: 12px;
  font-weight: bold;
  color: #95999D;
  text-transform: uppercase;
  letter-spacing: 0.6px;
}
.account h4 {
  font-size: 12px;
  font-weight: normal;
  color: #95999D;
  text-transform: uppercase;
  letter-spacing: 0.6px;
}
/* amount */
.amount {
  white-space: nowrap;
  font-size: 16px;
}
.amountPos {
  color: #35CE68;
}
.amountNeg {
  color: #F52D2D;
}
/* footer */
.footer__settings > td > div,
.footer__help > td > div {
  font-size: 12px;
  letter-spacing: 0.6px;
}

.footer__settings > td > div  {
  color: #32363F !important;
}
.footer__help > td > div {
  color: #95999D !important;
  text-align: right !important;
}
.footer__help a {
  display: inline-block;
  color: #95999D !important;
}

.footer__help a + a {
  margin-left: 16px;
}
/* text */
.txtRight {
  text-align: right;
}

/* media */
table {
  border-spacing: 0;
}

.tableRow {
  border-bottom: 1px #D6D8Da solid;
  padding: 8px 0;
  line-height: 1.5;
  letter-spacing: 0.25px;
}

.tableRow {
  border-bottom: 0;
}

.tableRow a, .tableRow a:visited {
  color: inherit;
  text-decoration: none;
}

@media screen and (max-width: 480px) {
  .header__desc__title {
    text-align: center !important;
  }
  .header__desc__subtitle {
    text-align: center !important;
  }
  .header__title > td {
    text-align: center !important;
  }
  .header__title > td > table {
    width: 100px;
    margin: auto;
  }

  .footer__settings > td > div {
    text-align: center !important;
  }

  .footer__help > td > div,
  td.footer__help > div,
  emailHeader {
    text-align: center !important;
  }
  .header__title table {
    margin: 0 auto;
    text-align: center;
  }
}

      </mj-style>
      <mj-attributes>
        <mj-all font-family=\\"Lato, Helvetica, Arial, sans-serif\\"></mj-all>
      </mj-attributes>
   </mj-head>
  <mj-body background-color=\\"#f5f6f7\\">
    <mj-section padding-top=\\"0\\" padding-bottom=\\"0\\">
      <mj-column>
            <mj-image width=\\"100px\\" src=\\"https://downcloud.cozycloud.cc/upload/cozy-banks/email-assets/logo-cozy.png\\" />

      </mj-column>
    </mj-section>

    <!-- white sheet -->
    <mj-wrapper background-color=\\"white\\" border='1px solid #d6d6d8' border-top=\\"none\\">
      <!-- header -->
      <mj-section padding-bottom=\\"0\\" padding-top=\\"0\\">
        <mj-column>
          <mj-table width=\\"100%\\" css-class=\\"header__title\\" cellpadding='4px'>
            <tr>
              <td width=\\"30px\\">
                <a href=\\"\\">
                  <img align=\\"left\\" width=\\"30px\\" src=\\"    https://downcloud.cozycloud.cc/upload/cozy-banks/email-assets/logo-my-app.png
\\" alt=\\"    My app
\\" />
                </a>
              </td>
              <td>
                <h1 style=\\"line-height: 32px\\"><a href=\\"\\">    My app
</a></h1>
              </td>
            </tr>
          </mj-table>
        </mj-column>
        <mj-column css-class=\\"emailHeader\\">
          <mj-text padding-top=\\"0\\" padding-bottom=\\"0\\">
              <div class=\\"header__desc__title txtRight\\">
                    My email title

              </div>
              <div class=\\"header__desc__subtitle txtRight\\">
                    Hello Homer !

              </div>
          </mj-text>
        </mj-column>
      </mj-section>

      <mj-section css-class='before-content' padding-top=\\"8px\\" padding-bottom=\\"8px\\">
        <mj-column>
          <mj-divider border-width=\\"1px\\" border-color=\\"#95999d\\" />
        </mj-column>
      </mj-section>


      <!-- content -->
          <mj-section>
      <mj-column>
        <mj-text>
          <h1>Title 1</h1>
          <h2>Title 2</h2>
          <p>Paragraph 1</p>
        </mj-text>
      </mj-column>
    </mj-section>
    <mj-section>
      <mj-column>
        <mj-text><p>Paragraph 2</p></mj-text>
      </mj-column>
    </mj-section>

    </mj-wrapper>

    <!-- footer -->
    <mj-section padding-top=\\"0\\">
      <mj-column>
        
      </mj-column>
      <mj-column>
        <mj-text align='right' css-class=\\"footer__help\\">
              <a href=\\"https://homer-myapp.mycozy.cloud\\">Go to app</a> -
    <a href=\\"https://homer-settings.mycozy.cloud\\">Go to settings</a>

        </mj-text>
      </mj-column>
    </mj-section>
  </mj-body>
</mjml>



"
`;
